src/Compilers/CSharp/Portable/FlowAnalysis/UnassignedFieldsWalker.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed class UnassignedFieldsWalker : DataFlowPass
+    {
+        private readonly DiagnosticBag _diagnostics;
+
+        private UnassignedFieldsWalker(CSharpCompilation compilation, MethodSymbol method, BoundNode node, DiagnosticBag diagnostics)
+            : base(compilation, method, node, trackClasses: true)
+        {
+            _diagnostics = diagnostics;
+        }
+
+        internal static void Analyze(CSharpCompilation compilation, MethodSymbol method, BoundNode node, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(method.MethodKind == MethodKind.Constructor);
+
+            var flags = ((CSharpParseOptions)node.SyntaxTree.Options).GetNullableReferenceFlags();
+            if ((flags & NullableReferenceFlags.Enabled) == 0)
+            {
+                return;
+            }
+
+            var walker = new UnassignedFieldsWalker(compilation, method, node, diagnostics);
+            try
+            {
+                bool badRegion = false;
+                walker.Analyze(ref badRegion, diagnostics: null);
+            }
+            finally
+            {
+                walker.Free();
+            }
+        }
+
+        protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
+        {
+            var result = base.Scan(ref badRegion);
+            ReportUninitializedNonNullableReferenceTypeFields();
+            return result;
+        }
+
+        private void ReportUninitializedNonNullableReferenceTypeFields()
+        {
+            var thisParameter = MethodThisParameter;
+            var location = thisParameter.ContainingSymbol.Locations.FirstOrDefault() ?? Location.None;
+
+            int thisSlot = VariableSlot(thisParameter);
+            if (thisSlot == -1)
+            {
+                return;
+            }
+
+            var thisType = thisParameter.Type.TypeSymbol;
+            Debug.Assert(thisType.IsDefinition);
+
+            foreach (var member in thisType.GetMembersUnordered())
+            {
+                if (member.Kind != SymbolKind.Field)
+                {
+                    continue;
+                }
+                var field = (FieldSymbol)member;
+                if (field.IsStatic || HasInitializer(field))
```

Comment:
>HasInitializer(field) [](start = 38, length = 21)

I am not sure why do we need to rely on presence of initializers. It feels like we should simply walk the body that already includes appropriate assignments.

Reply:
Added `// PROTOTYPE` comment.